### PR TITLE
Improve help message

### PR DESCRIPTION
**Pull request by leandro-lucarella-sociomantic**
There are still things to improve to make the CLI more consistent, like using always `-` or `--` for options, or using always `-option=xxx` for options taking an argument, right now the state is quite mixed and confusing.

### CODE REVIEW

**Comment by yebblies** [comment]
No implicit string literal concatenation!

**Comment by leandro-lucarella-sociomantic** [comment]
> No implicit string literal concatenation!

I tried to, but I got lost in compiler errors about conversion problems between pointers, immutable arrays and all kind of crazy things. I guess the intermediate stage between C++ and D code is not helping.

I'm open to suggestions on how to fix this though.

**Comment by leandro-lucarella-sociomantic** [comment]
Also, why not implicit concatenation? I remember discussions about wanting to remove it, but `~` is not the same (as it was proven by my failed attempts). I think is very useful to be able to split a string literal and for it to still be a string literal, and not an expression.

**Comment by WalterBright** [comment]
There is inconsistent use of plurals (i.e. `args`), ..., < > and [ ]. Some of the inconsistency was already there. But I suggest some consistency:

* `arg` to denote the text "arg"
* `<arg>` to denote an argument
* `{ <arg> }` to denote `<arg>` can appear 0 or more times
* `[ <arg> ]` to denote `<arg>` can appear 0 or one time

**Comment by WalterBright** [comment]
> I tried to, but I got lost in compiler errors about conversion problems between pointers, immutable arrays and all kind of crazy things.

Time for a bugzilla report!

**Comment by leandro-lucarella-sociomantic** [comment]
> I tried to, but I got lost in compiler errors about conversion problems between pointers, immutable arrays and all kind of crazy things.
> 
> Time for a bugzilla report!

I think is plain newbieness in terms of constness from my side (amplified by hard to read error-messages because of the unusually long string literal printed in the error). I am being just a tiny bit lazy here.

Let's try with a simpler example... Well, the error is at least suspicious:
```d
void f()
{
    const(char)* s = "some literal";
    auto ss = "hello" ~ s;
}
// Output: /tmp/s.d(6): Error: incompatible types for (("hello") ~ (s)): 'const(char)*' and 'const(char)*'
```
I guess this means that the operator `~` is not supported for `const(char)*` types, right? Is it bugzilla-worthy?

**Comment by leandro-lucarella-sociomantic** [comment]
> { <arg> } to denote <arg> can appear 0 or more times
> [ <arg> ] to denote <arg> can appear 0 or one time

At least in unixland, I never saw the form `{ <arg> }`, I only saw `[<arg>]` to denote zero or one (aka optional) and something like `[<arg> ...]` (or similar variants, like `[<arg ...>]`) to denote zero or more (and things like `<arg> [<arg> ...]` or `<arg ...>` to denote one or more).

**Comment by leandro-lucarella-sociomantic** [comment]
But to summarize, I prefer to avoid `{}` because I never saw it before so for me is very unclear. About any other variants, I just don't care, for me is just bikeshedding (they all look clear enough for me, with or without plurals, as long as the notation is consistent). Just let me know how do you prefer it and I'll change it.

**Comment by leandro-lucarella-sociomantic** [comment]
> No implicit string literal concatenation!

I could remove the literal concatenation by using something like this:
```d
auto s = "literal";
printf(("text" ~ "literal").ptr);
```

Is this acceptable or what would you recommend? I still think literal concatenation is much better to... concatenate literals, but well, I just want to get this merged to improve the first 5 minutes ;)

**Comment by leandro-lucarella-sociomantic** [comment]
> There is inconsistent use of plurals (i.e. args), ..., < > and [ ].

BTW, I don't see the inconsistency, could you be more specific? I used:

* `<files ...>` to denote 1 or more (plural + `...` is 1 or more)
* `<file>` to denote 1 (singular without `...` is strictly 1 mandatory thing)
* `[<arg>]` to denote one optional argument, 0 or 1 (`[ ]` meaning optional, and `<arg>` meaning the same as above)
* `[<args ...>]` to denote "optionally 1 or more" (which effectively is 0 or more)

Maybe you don't find this nomenclature clear enough but unless I'm missing something, I don't see inconsistencies...

**Comment by leandro-lucarella-sociomantic** [comment]
See `ls --help` or `git pus -h` for example. They all use nomenclatures very similar to what I'm using.

`ls`:
```
Usage: ls [OPTION]... [FILE]...
...
      --block-size=SIZE      scale sizes by SIZE before printing them.  E.g.,
      --color[=WHEN]         colorize the output.  WHEN defaults to 'always'
```

`git`:
```
usage: git push [<options>] [<repository> [<refspec>...]]
...
    --repo <repository>   repository
    --force-with-lease[=<refname>:<expect>]
```

**Comment by yebblies** [comment]
> I guess this means that the operator ~ is not supported for `const(char)*` types, right? Is it bugzilla-worthy?

No, it's behaving correctly.  It should work if those variables are changed to `enum string`s instead of `const(char)*`.

**Comment by leandro-lucarella-sociomantic** [comment]
I still would appreciate an explanation about why using literals concatenation is a bad idea :)

**Comment by yebblies** [comment]
Because implicit string literal concatenation causes bugs and we will hopefully get rid of it on day, so it should not be used in new code.

**Comment by leandro-lucarella-sociomantic** [comment]
> Because implicit string literal concatenation causes bugs

I don't mean to be a :trollface:, but causes bugs how exactly? Really, I'm trying to understand what's wrong with them :)

**Comment by yebblies** [comment]
```d
string[] x =
[
    "can",
    "you",
    "spot"
    "the",
    "bug",
    "?"
];
```

Couple that with the fact that we don't need them for anything, and...

**Comment by mathias-lang-sociomantic** [comment]
@yebblies : Maybe you could update https://dlang.org/deprecate.html then ? :)

**Comment by yebblies** [comment]
Chances are good I will if it ever gets implemented, since I wrote 90% of that page.

**Comment by mathias-lang-sociomantic** [comment]
> Chances are good I will if it ever gets implemented, since I wrote 90% of that page.

It's still useful as a mean to document future changes to the language, or guidelines.
Having stuff documented in people's mind doesn't scale up.
It can also provides prospective contributor with things to act on.

**Comment by leandro-lucarella-sociomantic** [comment]
OK, I just tried using `~` with the literals and worked. I think I got lost in the trial of different alternatives and actually did try that one.

Updated to use `~` to concatenate literals.

**Comment by WalterBright** [comment]
> I guess this means that the operator  ~  is not supported for  const(char)*  types, right?

Right.

> Is it bugzilla-worthy?

No, but the incorrect and unhelpful error message is bugzilla worthy.

**Comment by WalterBright** [comment]
@leandro-lucarella-sociomantic the git format is consistent and makes sense - let's go with that

**Comment by leandro-lucarella-sociomantic** [comment]
> @leandro-lucarella-sociomantic the git format is consistent and makes sense - let's go with that

So all I should change is plural -> singular and `<text ...>` -> `<text>...`, right? Or do you see anything else in my patch that's different from git's nomenclature?

**Comment by leandro-lucarella-sociomantic** [comment]
Updated to be closer to git's nomenclature. Also normalized all the options taking a filename to use `<filename>` and `<directory>` for the ones expecting a directory.

**Comment by wilzbach** [comment]
Btw if you are already doing this nice cleanup, can't we generate from this the man page?

**Comment by leandro-lucarella-sociomantic** [comment]
> Btw if you are already doing this nice cleanup, can't we generate from this the man page?

The only ways I see to make it happen, would involve a considerable amount of work, so I would leave it for a separate PR.

File: src/mars.d [hunk 1]
```diff
@@ -91,77 +91,84 @@ private  void usage()
 Documentation: http://dlang.org/
 Config file: %s
 Usage:
-  dmd files.d ... { -switch }
+  dmd [OPTIONS] <file>...
```
Comment (WalterBright) [inline src/mars.d]:
```
[<option>...]
```

File: src/mars.d [hunk 2]
```diff
@@ -91,77 +91,84 @@ private  void usage()
...
 
-  files.d        D source files
-  @cmdfile       read arguments from cmdfile
-  -allinst       generate code for all template instantiations
-  -betterC       omit generating some runtime information and helper functions
+Where:
+  <file>           D source file
+  <arg>            Argument to pass when running the resulting program
+
+OPTIONS:
```
Comment (WalterBright) [inline src/mars.d]: <option>:

File: src/mars.d [hunk 3]
```diff
@@ -91,77 +91,84 @@ private  void usage()
...
-  -vgc           list all gc allocations including hidden ones
-  -vtls          list all variables going into thread local storage
-  --version      print compiler version and exit
-  -version=level compile in version code >= level
-  -version=ident compile in version code identified by ident
-  -w             warnings as errors (compilation will halt)
-  -wi            warnings as messages (compilation will continue)
-  -X             generate JSON file
-  -Xffilename    write JSON file to filename
+  -c               do not link
```
Comment (WalterBright) [inline src/mars.d]:
```
-color    turn colored console output on
```

File: src/mars.d [hunk 4]
```diff
@@ -91,77 +91,84 @@ private  void usage()
...
+  -color[=on|off]  force colored console output on or off
```
Comment (WalterBright) [inline src/mars.d]:
```
=[on|off]
```

Comment (leandro-lucarella-sociomantic) [inline src/mars.d]: Mmm, this seems like a separate fix, but OK, I'll do it in a separate commit.

**Comment by leandro-lucarella-sociomantic** [comment]
Updated fixing @WalterBright comments.

**Comment by leandro-lucarella-sociomantic** [comment]
> No, but the incorrect and unhelpful error message is bugzilla worthy.

https://issues.dlang.org/show_bug.cgi?id=16097

**Comment by leandro-lucarella-sociomantic** [comment]
:tada: Thanks!

**Pull request merged**